pkg/api/methodHandlers.go

```diff
@@ -70,41 +72,46 @@ func NewMethodHandlers(adapterManager *server.AdapterManager, configManager *ser
 	}
 }
 
-func (h *methodHandlers) Check(ctx context.Context, tracker attribute.Tracker, request *mixerpb.CheckRequest, response *mixerpb.CheckResponse) {
-	// Prepare common response fields.
-	response.RequestIndex = request.RequestIndex
+type workFunc func(context.Context, attribute.MutableBag) *status.Status
 
-	ab, err := tracker.Update(request.AttributeUpdate)
+// does the standard attribute dance for each request
+func wrapper(ctx context.Context, tracker attribute.Tracker, attrs *mixerpb.Attributes, workFn workFunc) *status.Status {
+	ab, err := tracker.StartRequest(attrs)
 	if err != nil {
 		glog.Warningf("Unable to process attribute update. error: '%v'", err)
-		response.Result = newStatus(code.Code_INVALID_ARGUMENT)
-		return
+		return newStatus(code.Code_INVALID_ARGUMENT)
 	}
+	defer tracker.EndRequest()
 
 	// get a new context with the attribute bag attached
 	ctx = attribute.NewContext(ctx, ab)
 
-	dispatchKey, err := server.NewDispatchKey(ab)
-	if err != nil {
-		glog.Warningf("Error extracting the dispatch key. error: '%v'", err)
-		response.Result = newStatus(code.Code_FAILED_PRECONDITION)
-		return
-	}
+	return workFn(ctx, ab)
+}
 
-	allowed, err := h.checkLists(dispatchKey, tracker)
-	if err != nil {
-		glog.Warningf("Unexpected check error. dispatchKey: '%v', error: '%v'", dispatchKey, err)
-		response.Result = newStatus(code.Code_INTERNAL)
-		return
-	}
+func (h *methodHandlers) Check(ctx context.Context, tracker attribute.Tracker, request *mixerpb.CheckRequest, response *mixerpb.CheckResponse) {
+	response.RequestIndex = request.RequestIndex
+	response.Result = wrapper(ctx, tracker, request.AttributeUpdate,
+		func(ctx context.Context, ab attribute.MutableBag) *status.Status {
```

Comment:
I don't understand what you're suggesting. As it is, the callback for the wrapper function is provided inline. I could move the code to a separate function but that would just add lines with no real value.

As written, we don't currently close over any state from the parent.